cmd/crank/completion/completion.go

```diff
@@ -0,0 +1,259 @@
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	errFmtResourceTypeNotFound = "the server doesn't have a resource type %q"
+)
+
+// Predictors returns all supported predictors.
+func Predictors() map[string]complete.Predictor {
+	yamlPredictor := complete.PredictOr(
+		complete.PredictFiles("*.yml"),
+		complete.PredictFiles("*.yaml"),
+	)
+	return map[string]complete.Predictor{
+		"file":                   complete.PredictFiles("*"),
+		"xpkg_file":              complete.PredictFiles("*.xpkg"),
+		"yaml_file":              yamlPredictor,
+		"directory":              complete.PredictDirs("*"),
+		"file_or_directory":      complete.PredictOr(complete.PredictFiles("*"), complete.PredictDirs("*")),
+		"yaml_file_or_directory": complete.PredictOr(yamlPredictor, complete.PredictDirs("*")),
+		"namespace":              namespacePredictor(),
+		"context":                contextPredictor(),
+		"k8s_resource":           kubernetesResourcePredictor(),
+		"k8s_resource_name":      kubernetesResourceNamePredictor(),
+	}
+}
+
+func kubernetesResourcePredictor() complete.PredictFunc {
+	return func(a complete.Args) []string {
+		_, kubeconfig, err := kubernetesClient(parseConfigOverride(a))
+		if err != nil {
+			return nil
+		}
+
+		discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeconfig)
+		if err != nil {
+			return nil
+		}
+
+		resources, err := discoveryClient.ServerPreferredResources()
+		if err != nil {
+			return nil
+		}
+
+		if len(resources) == 0 {
+			return nil
+		}
+
+		var predictions []string
+		for _, apiResources := range resources {
+			for _, res := range apiResources.APIResources {
+				var resourceName string
+
+				switch {
+				case res.Group != "" && res.Version != "":
+					resourceName = fmt.Sprintf("%s.%s.%s", res.Name, res.Version, res.Group)
+				case res.Version != "":
+					resourceName = fmt.Sprintf("%s.%s", res.Name, res.Version)
+				default:
+					resourceName = res.Name
+				}
+
+				if strings.HasPrefix(resourceName, a.Last) {
+					predictions = append(predictions, resourceName)
+				}
+			}
+		}
+		return predictions
+	}
+}
+
+func kubernetesResourceNamePredictor() complete.PredictFunc {
+	return func(a complete.Args) []string {
+		client, kubeconfig, err := kubernetesClient(parseConfigOverride(a))
+		if err != nil {
+			return nil
+		}
+
+		discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeconfig)
+		if err != nil {
+			return nil
+		}
+
+		d := memory.NewMemCacheClient(discoveryClient)
+		rmapper := restmapper.NewShortcutExpander(restmapper.NewDeferredDiscoveryRESTMapper(d), d, nil)
+		mapping, err := mappingFor(rmapper, a.LastCompleted)
+		if err != nil {
+			return nil
+		}
+
+		// Get Resource object. Contains k8s resource and all its children, also as Resource.
```

Comment:
i think this comment was copied from the `trace` command and it doesn't make sense here - this list isn't going to recursively search for children.